spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jdbc/HikariLifecycle.java

```diff
@@ -0,0 +1,221 @@
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+
+import com.zaxxer.hikari.HikariDataSource;
+import com.zaxxer.hikari.HikariPoolMXBean;
+import com.zaxxer.hikari.pool.HikariPool;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.aop.TargetSource;
+import org.springframework.aop.framework.Advised;
+import org.springframework.aop.support.AopUtils;
+import org.springframework.context.Lifecycle;
+import org.springframework.util.Assert;
+import org.springframework.util.ReflectionUtils;
+
+/**
+ * {@link Lifecycle} management for a {@link HikariDataSource} taking care of
+ * {@link Lifecycle#start() starting}/{@link Lifecycle#stop() stopping} the
+ * {@link javax.sql.DataSource} by {@link HikariDataSource#isAllowPoolSuspension()
+ * eventually} suspending/resuming the underlying {@link HikariPool connection pool} and
+ * {@link HikariPoolMXBean#softEvictConnections() evicting} open & idle connections.
+ *
+ * @author Christoph Strobl
+ */
+class HikariLifecycle implements Lifecycle {
+
+	private final HikariDataSource dataSource;
+
+	private final LifecycleExecutor lifecycleExecutor;
+
+	HikariLifecycle(HikariDataSource dataSource) {
+
+		this.dataSource = dataSource;
+		this.lifecycleExecutor = new LifecycleExecutor(dataSource);
+	}
+
+	HikariDataSource getManagedInstance() {
+		return this.dataSource;
+	}
+
+	@Override
+	public void start() {
+
+		if (this.dataSource.isRunning()) {
+			return;
+		}
+
+		if (this.dataSource.isClosed()) {
+			throw new IllegalStateException("DataSource has been closed and cannot be restarted");
+		}
+
+		this.lifecycleExecutor.resume();
+	}
+
+	@Override
+	public void stop() {
+		if (this.dataSource.isRunning()) {
+			this.lifecycleExecutor.pause();
+		}
+	}
+
+	@Override
+	public boolean isRunning() {
+		return this.dataSource.isRunning();
+	}
+
+	/**
+	 * Component to help suspend/resume a {@link HikariDataSource} by taking the pool
+	 * suspension flag into account. Will perform best effort to make sure connections
+	 * reported as closed buy the {@link HikariPoolMXBean} have actually been closed by
+	 * the {@link java.util.concurrent.Executor} that is in charge of closing them.
+	 *
+	 * @author Christoph Strobl
+	 */
+	private static class LifecycleExecutor {
+
+		private static final Log logger = LogFactory.getLog(LifecycleExecutor.class);
+
+		private final HikariDataSource dataSource;
+
+		private Supplier<Boolean> hasOpenConnections;
+
+		LifecycleExecutor(HikariDataSource hikariDataSource) {
+
+			this.dataSource = getUltimateTargetObject(hikariDataSource);
+
+			if (hikariDataSource.getHikariPoolMXBean() instanceof HikariPool pool) {
+
+				Field closeConnectionExecutor = ReflectionUtils.findField(HikariPool.class, "closeConnectionExecutor");
+				if (closeConnectionExecutor != null) {
```

Comment:
If the field's not found something's gone really wrong. I think it would be better to fail hard.

Reply:
The idea was to be very defensive here, and fall back to checking _total connections_ if for whatever reason `HikariPool` does not meet the expectations.

Reply:
If Hikari doesn't meet expectations, then I think we should just give up as we have no way of knowing that our approach is valid. We'll know, through tests, that it works with the version of Hikari in dependency management. If someone overrides that version and Hikari has changed incompatibility I think we'd be better to fail fast. That'll make the problem much easier to diagnose rather than dealing with a checkpoint failure and trying to figure out why it occurred.